src/services/hsm.service.js

```diff
@@ -0,0 +1,203 @@
+    host: config.hsmHost,
+    authUsernamePassword: {
+      username: config.hsmUsername,
+      password: config.hsmPassword,
+    },
+  };
+
+  // Connects to HSM and returns the connection object
+  return hsm.connect(hsmConnectionOptions);
+}
+
+/**
+ * Disconnects the headless wallet from the HSM.
+ * The connection object is a singleton, so it must not be informed as parameter.
+ * @returns {Promise<void>}
+ */
+async function hsmDisconnect() {
+  await hsm.disconnect();
+  lock.unlock(lockTypes.HSM);
+}
+
+/**
+ * Checks if the informed HSM key name contains a valid BIP32 xPriv key
+ * @param {Object} hsmConnection
+ * @param {string} hsmKeyName
+ * @returns {Promise<{reason?: string, isValidXpriv: boolean, hsmKeyName}>}
+ */
+async function isKeyValidXpriv(hsmConnection, hsmKeyName) {
+  const returnObject = {
+    hsmKeyName,
+    isValidXpriv: false,
+    reason: '',
+  };
+
+  // Checks if the key exists on HSM
+  let keyInfo;
+  try {
+    keyInfo = await hsmConnection.blockchain.getKeyInfo(hsmKeyName);
+  } catch (e) {
+    // Special treatment for a wallet not configured on the HSM
+    returnObject.reason = e.errorCode === 5004
+      ? `Key does not exist on HSM.`
+      : `Unexpected error on key retrieval: ${e.message}`;
+    return returnObject;
+  }
+
+  // Validates the key type and version
+  if (keyInfo.type !== hsm.enums.BLOCKCHAIN_KEYS.BIP32_XPRV) {
+    returnObject.reason = `hsm-key is not a valid xPriv.`;
+    return returnObject;
+  }
+  if (!(keyInfo.ver in [
+    hsm.enums.VERSION_OPTIONS.BIP32_HTR_MAIN_NET,
+    hsm.enums.VERSION_OPTIONS.BIP32_HTR_TEST_NET,
+  ])) {
+    returnObject.reason = `hsm-key is not a valid Hathor xPriv.`;
+    return returnObject;
+  }
+
+  // Returns a positive result for a valid xPriv key
+  delete returnObject.reason;
+  returnObject.isValidXpriv = true;
+  return returnObject;
+}
+
+/**
+ * Derivates an HTR xPriv key from a given xPriv key
+ * @param {Object} hsmConnection
+ * @param {string} hsmKeyName
+ * @param {Object} [options]
+ * @param {boolean} [options.isReadOnlyWallet] If true, derivation will stop at the account level
+ * @param {number} [options.version] Optional, advanced. Version to calculate the exported xPriv
+ * @returns {Promise<{success: boolean, htrKeyName: string}>}
+ */
+async function derivateHtrCkd(
+  hsmConnection,
+  hsmKeyName,
+  options,
+) {
+  const childKeyNames = {
+    HTR_CKD_BIP_KEYNAME: 'HTR_BIP_KEY',
+    HTR_CKD_COIN_KEYNAME: 'HTR_COIN_KEY',
+    HTR_CKD_ACCOUNT_KEYNAME: 'HTR_ACCOUNT_KEY',
+    HTR_CKD_CHANGE_KEYNAME: 'HTR_CHANGE_KEY',
+    HTR_CKD_ADDRESS_KEYNAME: 'HTR_ADDRESS_KEY',
+  };
+
+  /*
+   * Derivation will be made on
+   * BIP code / Coin / Account /  Change
+   *    m/44' / 280' /      0' /       0
+   *        1 /    2 /      3  /       4
+   */
+
+  // Derivation 1: Bip Code
+  const derivationVersion = options?.version
+    ? options.version
+    : hsm.enums.VERSION_OPTIONS.BIP32_HTR_MAIN_NET;
```

Comment:
Why don't you get the network in the config, so you can use the default as testnet here, in case this is the configuration?

Reply:
Added on [b332161](https://github.com/HathorNetwork/hathor-wallet-headless/pull/356/commits/b3321613176a720d172eb4f12898e34f345ad84e)